[ethernet][core] LLDP 802.1AB transmitter

Gnuspice opened the pull request:
# What does this implement/fix?

Adds LLDP (802.1AB) transmission support to the Ethernet component, so ethernet-capable ESPHome devices can report their details to managed ethernet switches.

With an `ethernet.lldp` section defined, this periodically sends LLDP packets over the ethernet link, enabling users with appropriately-featured switches to see what devices are plugged in where.

Tested against Linux "lldpd" version 1.0.18, a Juniper EX-series switch, and manual verification of generated packets using Wireshark.

When not enabled (no `ethernet.lldp` section), the code/binary are identical to builds not including this feature. When enabled, it adds approximately 9kb to the total image size (5.5kb code, 3.3kb data, some of which is the inclusion of the L2TAP ESP-IDF feature).

Support for receiving and re-publishing LLDP frames as sensor values is not included in this patch.

Real-world functionality confirmed on:

* M5Stack PoESP32 device with IP101 ethernet chipset
* Seeed Studio XIAO W5500 device with W5500 ethernet chipset
* Olimex ESP32-POE-ISO board with LAN8720 ethernet chipset

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5573

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
ethernet:
  # [... ethernet device configuration ...]
  lldp:
    # Optional: override default port
    port: "test-port"
    # Optional: override default system name
    system_name: "test-system"
    # Optional: override default system description
    system_description: "test-description"
    # Optional: override default "fast mode" burst packet count
    tx_fast_count: 4
    # Optional: override default LLDP frame transmission interval
    tx_interval: 30
    # Optional: override default hold count, used in conjuction with tx_interval to calculate overall TTL
    tx_hold: 4

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Gnuspice committed:
Ethernet LLDP 802.1AB transmitter

Add LLDP transmission support to the Ethernet component, so
ethernet-capable ESPHome devices can report their details to managed
ethernet switches.

bdraco commented on esphome/components/ethernet/ethernet_component.h:
```diff
@@ -114,6 +126,21 @@ class EthernetComponent : public Component {
+  uint16_t use_lldp_tx_hold_;
+  std::string use_lldp_port_;
+  std::string use_lldp_name_;
+  std::string use_lldp_desc_;
```
Anything set from python codegen should be `const char *` so its stored in flash and not ram

Gnuspice replied:
Updated to use `const char *` throughout in latest commit - thanks.

bdraco commented on esphome/components/ethernet/lldp.h:
```diff
@@ -0,0 +1,146 @@
+  network::IPAddresses ip_addresses_;
+  std::string port_;
+  std::string system_name_;
+  std::string system_description_;
```
Please store strings in flash

Gnuspice replied:
Updated to use `const char *` throughout in latest commit - thanks.

Gnuspice committed:
LLDP: change from std::string to const char *

Updated as per code review comments on
<https://github.com/esphome/esphome/pull/11760>

jesserockz commented on esphome/components/ethernet/__init__.py:
```diff
@@ -145,6 +154,18 @@
 
+LLDP_SCHEMA = cv.Schema(
+    {
+        cv.Optional(CONF_ENABLED, default=False): cv.boolean,
```
Is there a case when you would configure `lldp` in yaml but want it disabled? If so that is fine, but we can probably default it to True as we can assume you do want it enabled if you are configuring it.
```suggestion
        cv.Optional(CONF_ENABLED, default=True): cv.boolean,
```

Gnuspice replied:
Not really, no - the specific reason for `ethernet.lldp.enabled` existing was so the LLDP transmitter could be enabled without modifying any of the default values, and without the ugliness of`lldp: {}` YAML.

I'd had a look around the code where there are examples handling both:

`feature: true`

and:

```
feature:
  key: val
  key2: val2
```

but it all looked rather ad-hoc, so I didn't want to go there without guidance.

Your suggestion makes `lldp: {}` work as would more reasonably be expected - LLDP on, with default settings - happy to make that change if a more optimal solution doesn't come to mind.

Gnuspice replied:
Made the default=True change, then thought more about it and did away with `enabled` entirely.

jesserockz commented on esphome/components/ethernet/__init__.py:
```diff
@@ -403,6 +425,27 @@ async def to_code(config):
         # Add LAN867x 10BASE-T1S PHY support component
         add_idf_component(name="espressif/lan867x", ref="2.0.0")
 
+    if CONF_LLDP in config and config[CONF_LLDP][CONF_ENABLED]:
```
```suggestion
    if (lldp_config := config.get(CONF_LLDP)) and lldp_config[CONF_ENABLED]:
```

Gnuspice replied:
Done, and also removed the `enabled` portion.

jesserockz commented on esphome/components/ethernet/__init__.py:
```diff
@@ -403,6 +425,27 @@ async def to_code(config):
+        if system_name := config[CONF_LLDP].get(CONF_SYSTEM_NAME):
+            cg.add(var.set_lldp_system_name(system_name))
+        if system_description := config[CONF_LLDP].get(CONF_SYSTEM_DESCRIPTION):
+            cg.add(var.set_lldp_system_description(system_description))
```
```suggestion
        cg.add(var.set_lldp_tx_fast_count(lldp_config[CONF_TX_FAST_COUNT]))
        cg.add(var.set_lldp_tx_interval(lldp_config[CONF_TX_INTERVAL]))
        cg.add(var.set_lldp_tx_hold(lldp_config[CONF_TX_HOLD]))
        if port := lldp_config.get(CONF_PORT):
            cg.add(var.set_lldp_port(port))
        if system_name := lldp_config.get(CONF_SYSTEM_NAME):
            cg.add(var.set_lldp_system_name(system_name))
        if system_description := lldp_config.get(CONF_SYSTEM_DESCRIPTION):
            cg.add(var.set_lldp_system_description(system_description))
```

Gnuspice replied:
Done.

jesserockz reviewed (changes requested):
Have yet to look at cpp, this is only a review of the python side for now

Gnuspice committed:
LLDP: change lldp.enabled default, tidy codegen

Updated lldp.enabled default value from False to True and tidied up
codegen section as per code review comments.

Gnuspice committed:
LLDP: remove lldp.enabled flag

Do away with the `enabled` boolean - just enable the transmitter
whenever an `ethernet.lldp` section is defined.

Gnuspice commented:
Updated with changes requested; upon further thought, ended up removing the `ethernet.lldp.enabled` setting entirely; enabling with default settings is just now `lldp: {}`

Docs PR updated to match.

Gnuspice committed:
LLDP: remove extraneous ESP_LOGCONFIG call

Combine two ESP_LOGCONFIG() calls into one.

martijnbuts commented:
Hi, tested the pr with a ESP32-P4 and a switch and it works. Thanks for all the work!

ssieb commented on esphome/components/ethernet/ethernet_component.cpp:
```diff
@@ -18,6 +18,20 @@
+#define LLDP_DEFAULT_DESCRIPTION "ESPHome.io " ESPHOME_VERSION " " ESPHOME_BOARD
+#endif
+
+#include "esphome/core/version.h"
```
This include needs to be higher.

Gnuspice replied:
This and other include issues you noted are addressed in the latest commit.

ssieb commented on esphome/components/ethernet/ethernet_component.cpp:
```diff
@@ -18,6 +18,20 @@
+#endif
+
+#include "esphome/core/version.h"
+#include "lldp.h"
```
You already included this in the `.h` file.

ssieb commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,409 @@
+// All mandatory and some optional data units have been implemented.
+//
+// Written by Nic Bellamy, 2025. https://github.com/Gnuspice
+#include "esphome/core/defines.h"
```
Already in the `.h` file.

Gnuspice committed:
LLDP: tidy up #includes

Tidy up #includes so headers aren't included more than necessary to
address review comments from @ssieb

Gnuspice committed:
Merge branch 'dev' into ethernet-lldp-support

Gnuspice commented:
Updated description to note successful testing on a Seeed Studio XIAO W5500 module (ESP32-S3 with W5500 ethernet, see <https://wiki.seeedstudio.com/xiao_w5500_ethernet_adapter/>).

```yaml
# Configuration excerpt for LLDP on Seed Studio W5500 PoE ethernet module
ethernet:
  type: W5500
  clk_pin: GPIO7
  mosi_pin: GPIO9
  miso_pin: GPIO8
  cs_pin: GPIO2
  lldp: {}
```

swoboda1337 committed:
Merge remote-tracking branch 'origin/dev' into ethernet-lldp-support

swoboda1337 committed:
Merge remote-tracking branch 'upstream/dev' into ethernet-lldp-support

# Conflicts:
#	esphome/components/ethernet/__init__.py
#	esphome/components/ethernet/ethernet_component.cpp

swoboda1337 commented:
No problem, I just got Claude to do it :)

randybb commented:
It was tested on Olimex poe-iso esp32 board with LAN8720.

Gnuspice committed:
Merge remote-tracking branch 'upstream/dev' into ethernet-lldp-support

Gnuspice committed:
[ethernet.lldp] Use C++17 nested namespace syntax

Gnuspice commented:
Merged to latest dev branch, converted to C++17 namespacing syntax, and ran it through its paces locally to ensure everything is still happy on real hardware.

bdraco committed:
Merge branch 'dev' into ethernet-lldp-support

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+#include "esp_eth.h"
+#include "esp_vfs_l2tap.h"
+
+#define SIZE_TLV_END (2)
```
Please use  `static constexpr` instead of define

bdraco commented on esphome/components/ethernet/lldp.h:
```diff
@@ -0,0 +1,144 @@
+#include "esphome/components/network/ip_address.h"
+#include "esp_eth.h"
+
+// Initial configuration
```
Please use  `static constexpr` instead of define

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+#define LLDP_ERROR_CHECK(ret) \
+  if (!ret) { \
+    return false; \
+  }
```
This looks like unneeded complexity

bdraco commented on esphome/components/ethernet/ethernet_component.h:
```diff
@@ -148,6 +160,21 @@ class EthernetComponent : public Component {
+  uint16_t use_lldp_tx_hold_;
+  const char *use_lldp_port_{""};
+  const char *use_lldp_name_{""};
+  const char *use_lldp_desc_{""};
```
`use_` for member vars is a bit unusual

Could just be `lldp_tx_fast_count_`; etc

bdraco commented on esphome/components/ethernet/ethernet_component.h:
```diff
@@ -103,6 +107,14 @@ class EthernetComponent : public Component {
+  void set_lldp_port(const char *use_port) { this->use_lldp_port_ = use_port; };
+  void set_lldp_system_name(const char *use_name) { this->use_lldp_name_ = use_name; };
+  void set_lldp_system_description(const char *use_desc) { this->use_lldp_desc_ = use_desc; };
+  void set_lldp_tx_fast_count(const uint16_t &use_count) { this->use_lldp_tx_fast_count_ = use_count; };
```
These are simple types, its a bit unusual to pass them by const ref

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+#include <string.h>
+#include <unistd.h>
+
+#include <string>
```
stale include

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+#ifdef USE_ETHERNET_LLDP
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
```
can be moved to .cpp

bdraco commented on esphome/components/ethernet/lldp.h:
```diff
@@ -0,0 +1,144 @@
+  const char *system_description_{""};
+  struct eth_addr lldp_multicast_mac_;
+  struct eth_addr mac_addr_;
+  int l2tap_fd_;
```
```suggestion
  int l2tap_fd_{-1};
```

bdraco commented on esphome/components/ethernet/lldp.h:
```diff
@@ -0,0 +1,144 @@
+  bool generate(uint8_t *buf, size_t buf_len, size_t *pkt_len);
+
+  // Recalculate TTL based on interval and hold values
+  void recalc_ttl() { this->calc_ttl_ = std::min(this->tx_interval_ * this->tx_hold_ + 1, LLDP_PROTO_MAX_TTL); };
```
this can overflow

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+      }
+      this->state_ = TXStateNormal;
+      this->state_message_interval_ = this->tx_interval_;
+      // FALLTHROUGH
```
[[fallthrough]];

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+  return ESP_OK;
+}
+
+void LLDPTransmitter::set_ips(network::IPAddresses &ips) {
```
```suggestion
void LLDPTransmitter::set_ips(const network::IPAddresses &ips) {
```

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+    // If the ethernet port is not connected (no link, maybe other cases), this
+    // will return errno 5 - EIO "I/O error". The file descriptor starts
+    // working again when the link is reestablished.
+    ESP_LOGE(TAG, "LLDP L2TAP write(%d, %d) error %d: %s", this->l2tap_fd_, pkt_len, errno, strerror(errno));
```
pkt_len %zu

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+    return ESP_FAIL;
+  }
+  if (ret != pkt_len) {
+    ESP_LOGE(TAG, "LLDP L2TAP short write(%d, %d): only wrote %d bytes", this->l2tap_fd_, pkt_len, ret);
```
pkt_len %zu

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+    return ESP_FAIL;
+  }
+
+  ESP_LOGD(TAG, "Sent %d byte LLDPDU successfully", pkt_len);
```
pkt_len %zu

bdraco commented on esphome/components/ethernet/lldp.h:
```diff
@@ -0,0 +1,144 @@
+//
+// All mandatory and some optional data units have been implemented.
+//
+// Written by Nic Bellamy, 2025. https://github.com/Gnuspice
```
Please use CODEOWNERS instead

Gnuspice replied:
What would be the appropriate way to do this? I don't want to add myself to `ethernet/__init__.py` as that'll mark me as owner of the ethernet component, which I'm not - and the `build_codeowners.py` script doesn't handle file-level ownership.

Making an `lldp` subdirectory with its own `__init__.py` doesn't seem quite right either, and won't do the right thing in CODEOWNERS unless I also move `lldp.h` and `lldp.cpp` in there. The subdirectory approach seems to be used for sensors/switches/etc on a base platform, rather than optional integrated features, in all cases I've looked at hunting around in the components directory. I'm not sure how to get it to even compile the lldp files in that case.

(Separately, is it against policy somewhere to have an attribution in the comment like I have? Happy to try and sort out CODEOWNERS, but would like to keep attribution somewhere that can't be revoked at-will by a third party eg. Github nuking my account. Perhaps reworded to something like "Originally written by Nic Bellamy, see CODEOWNERS for current maintainer"?)

bdraco replied:
We don’t currently have a formal policy around author attribution comments, and defining one just to move this PR forward would likely add unnecessary process. After discussing internally, we think there’s a better structural and architectural solution that we should implement regardless.

Since LLDP is really an optional feature, it makes more sense for it to be its own `lldp` component rather than living under `ethernet` that depends on `ethernet`. That way you can be the `CODEOWNER` of the `lldp` component specifically, without being marked as owner of the entire `ethernet` component.

Gnuspice replied:
I had considered this approach originally, eg. an `ethernet_lldp` component, but had decided against it as it'd need more invasive changes to the ethernet module - I need access to things that don't have existing architecture to handle. It does straightforwardly resolve the CODEOWNERS issue though, as you say.

I'm happy to explore this approach - but we'll have to figure out how to hook things, which is not something I wish to attempt without some guidance.

In particular, I'd need to be able to, from outside `ethernet`:

* Get EthernetComponent.eth_handle_, required for binding the L2TAP socket to the network interface (no API)
* Receive ethernet state changes (no API)
* Get current IPs - `EthernetComponent.get_ip_addresses()` already exists and should work
* Get MAC address - `EthernetComponent.get_eth_mac_address_raw()` already exists and should work

I'm currently doing LLDP setup at the end of `Ethernet::setup()`, but should be able to do initialisation on first connect event instead, rather than needing a special end-of-setup hook. It will just mean any LLDP setup error would be delayed until first connect.

bdraco replied:
> * Get EthernetComponent.eth_handle_, required for binding the L2TAP socket to the network interface (no API)

Probably go with the simple approach
`esp_eth_handle_t get_eth_handle() const { return this->eth_handle_; }`

bdraco replied:
> * Receive ethernet state changes (no API)

We already have
```
  #ifdef USE_ETHERNET_STATE_LISTENERS
  class EthernetStateListener {
   public:
    virtual void on_ethernet_connected() = 0;
    virtual void on_ethernet_disconnected() = 0;
  };
  #endif
```

Gnuspice replied:
What branch does USE_ETHERNET_STATE_LISTENERS live in? I'm looking at dev 656389f21 and can't see it.

It does look like a workable solution though.

Gnuspice replied:
I've done some playing around over the weekend, and have, with the addition of a `get_eth_handle()` function to EthernetComponent, a "mostly working" separate `ethernet_lldp` component, doing setup at `setup_priority::ETHERNET - 1.0f`, and using EthernetIPStateListener to get IP addressing updates.

Whether `EthernetStateListener` will complete the picture will depend on the exact semantics of `on_ethernet_connected()`, which I'm not yet privy to.

If it simply means "link up" (ie. triggers from an ESP-IDF `ETHERNET_EVENT_CONNECTED` event), it'll work nicely.

If it means "connected" in the sense `EthernetComponent.connected_` uses - "link up and have enough IP addresses to consider things ready to go" - then it won't, as from a protocol perspective, LLDP fast bursts should begin on link up not on IP assignment, and from a functionality perspective, LLDP being able to tell the switch "my link is up but I don't have a management address" is a useful part of it.

bdraco replied:
It would be fine to add another `on_ethernet_link_up` or whatever is needed to EthernetStateListener in another PR

Gnuspice replied:
You said above "We already have" the `EthernetStateListener` interface - but I can't find it anywhere.

```
~/src/esphome
$ git log --oneline -n1
19bbd39e3 (HEAD -> dev, upstream/dev) [uart] Enable wake-on-RX by default on ESP32 (#14391)
$ git grep EthernetStateListener
$ git grep USE_ETHERNET_STATE_LISTENERS
$ git grep on_ethernet_connected
$ git grep on_ethernet_disconnected
$
```

Github search only returns this ticket.

bdraco replied:
Not at my desk at the moment, but there should be an example in ethernet_info

bdraco replied:
It's also equally possible it's only on my integration branch and we never merged or PRed it. I'll check on Monday.

ssieb replied:
It must be something you only have locally.  I can't find it anywhere in the source or other PRs.

bdraco replied:
Yup its only in my local merge conflict hell. It was something I was going to add to ethernet info to replace https://github.com/esphome/esphome/pull/13203 with something like #14084 but I never finished the PR because the conflicts got too bad and I forgot about it

bdraco replied:
My local is too messy. Best path forward is to do a fresh PR before this one with the pattern in https://github.com/esphome/esphome/pull/14084

Gnuspice replied:
Righto, I'll have a crack at that, and do a separate small PR for the `get_eth_handle()` change.

There's some complexity on ethernet state callbacks as they'll need to get called with current state, or we'll end up with "ethernet init" then "LLDP init + callback registration" then no callbacks because the link started up prior to LLDP registering the callbacks. Details on how to do that probably best discussed on eventual ethernet state callbacks PR.

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+  // =============
+
+  // Send System Name if set
+  if (strlen(this->system_name_)) {
```
standard pattern is to init to `nullptr` and check if its not nullptr

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+  LLDPPacket pkt(buf, buf_len);
+
+  // Start the packet off by building an ethernet header
+  pkt.begin(this->mac_addr_, this->lldp_multicast_mac_);
```
return is ignored here

bdraco commented on esphome/components/ethernet/lldp.h:
```diff
@@ -0,0 +1,144 @@
+  const char *port_{""};
+  const char *system_name_{""};
+  const char *system_description_{""};
+  struct eth_addr lldp_multicast_mac_;
```
looks like this is always a const so doesn't. need to be a member var

bdraco commented on esphome/components/ethernet/ethernet_component.cpp:
```diff
@@ -734,6 +774,62 @@ void EthernetComponent::dump_connect_params_() {
+  }
+
+  // Set up 1-second tick handler for the LLDP transmit engine
+  this->set_interval("lldp_timer_tick", 1000, [this] { this->lldp_timer_tick_(); });
```
Looks like this never cancels when ethernet disconnects/restarts

Gnuspice replied:
Correct - it didn't seem necessary, as worst-case the tick function will emit a transmit error message occasionally (every 30 seconds by default) when disconnected, but will start working again when ethernet reconnects.

Happy to change if desired - I will just need to read through the state transitions carefully to ensure I catch everything.

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+    return err;
+  }
+
+  int fd = ::open(L2TAP_VFS_DEFAULT_PATH, 0);
```
0 flag but used for write ?

Gnuspice replied:
Should have been O_WRONLY - fixed.

bdraco commented on esphome/components/ethernet/__init__.py:
```diff
@@ -169,6 +177,17 @@ def request_ethernet_ip_state_listener() -> None:
+    {
+        cv.Optional(CONF_PORT): cv.string_strict,
+        cv.Optional(CONF_SYSTEM_NAME): cv.string_strict,
+        cv.Optional(CONF_SYSTEM_DESCRIPTION): cv.string_strict,
```
LLDP_STRING_MAX

Missing  cv.Length(max=128)

bdraco commented on esphome/components/ethernet/lldp.h:
```diff
@@ -0,0 +1,144 @@
+  int l2tap_fd_;
+  uint16_t calc_ttl_;
+  uint16_t state_fastmode_count_;
+  uint16_t state_message_interval_;
```
```suggestion
  uint16_t state_fastmode_count_{0};
  uint16_t state_message_interval_{0};
```

bdraco commented on esphome/components/ethernet/lldp.cpp:
```diff
@@ -0,0 +1,406 @@
+//
+// This needs to match up with the TLVs (and their sizes) used in LLDPTransmitter::generate()
+//
+const size_t PKT_MAX = sizeof(struct eth_hdr)                    /* eth_hdr */
```
constexpr

bdraco commented on esphome/components/ethernet/lldp.h:
```diff
@@ -0,0 +1,144 @@
+        tx_fast_count_{LLDP_TX_FAST_COUNT},
+        tx_interval_{LLDP_TX_INTERVAL},
+        tx_hold_{LLDP_TX_HOLD},
+        state_{TXStateReconnect} {
```
Looks like needs to be reordered to avoid  -Wreorder

bdraco commented on esphome/components/ethernet/__init__.py:
```diff
@@ -455,6 +496,9 @@ async def to_code(config):
         )
 
+    if CORE.using_arduino:
+        cg.add_library("WiFi", None)
```
Is this needed?

Gnuspice replied:
This wasn't my code, but had snuck back in when someone was resolving merge conflicts; removed.

bdraco commented on esphome/components/ethernet/__init__.py:
```diff
@@ -443,6 +463,27 @@ async def to_code(config):
+        add_idf_sdkconfig_option("CONFIG_ESP_NETIF_L2_TAP_MAX_FDS", 1)
+        add_idf_sdkconfig_option("CONFIG_ESP_NETIF_L2_TAP_RX_QUEUE_SIZE", 1)
+
+        cg.add_define("USE_ETHERNET_LLDP")
```
Please add to defines.h for static analysis

bdraco reviewed (changes requested):
Please see comments above

bdraco commented on esphome/components/ethernet/lldp.h:
```diff
@@ -0,0 +1,144 @@
+        state_{TXStateReconnect} {
+    this->recalc_ttl();
+  }
+
```
Never destroyed so ok

Gnuspice replied:
Seconded - marking resolved.

Gnuspice committed:
LLDP: address most @bdraco code review issues

* Switch from #define to constexpr
* Remove LLDP_ERROR_CHECK() macro and add missing return value check
* Remove `use_` prefix on ethernet component LLDP settings
* Remove const ref usage for basic type
* Remove stale includes
* Ensure l2tap_fd_ is initialized to -1 to protect against misuse
* Fix potential uint16_t overflow in TTL calculation
* Use modern `[[fallthough]]` tag in state engine switch{}
* Fix size_t printf format specifiers
* Use nullptr instead of an empty string for configurable string parameters
* Use the LLDP multicast MAC address directly rather than via a member variable
* Use correct O_WRONLY flag value on L2TAP open() call
* Initialise various member variables to protect against misuse
* Fix -Wreorder issues in LLDPTransmitter constructor
* Add USE_ETHERNET_LLDP to core/defines.h for static analysis as directed
* Add cv.Length(max=128) validators for port/name/description strings

Gnuspice committed:
LLDP: Remove old code from merge conflict

The following old code appears to have snuck back into the codebase
during resolution of merge conflicts in e991f30015471acfbb449b8fb257ca62f9893bf7

```
    if CORE.using_arduino:
        cg.add_library("WiFi", None)
```

Code was originally from commits
c2e198311ca73112fa7c9c213101277200085a6f and 1627dff166768ec79928fc3e65a3b18f1d7267d4

Gnuspice committed:
LLDP: Fix clang-tidy errors

Linters are annoying, especially when they're right.

Gnuspice commented on esphome/components/ethernet/__init__.py:
```diff
@@ -455,6 +496,9 @@ async def to_code(config):
         )
 
+    if CORE.using_arduino:
+        cg.add_library("WiFi", None)
```
This wasn't my code, but had snuck back in when someone was resolving merge conflicts; removed.

Gnuspice committed:
Merge branch 'dev' into ethernet-lldp-support

Gnuspice committed:
Merge branch 'dev' into ethernet-lldp-support
